contrib/utilities/prm_tools.py

```diff
@@ -0,0 +1,315 @@
+## LICENSE.md and CONTRIBUTING.md at the top level directory of deal.II.
+##
+## ------------------------------------------------------------------------
+
+""" Routines to read, format and write deal.II parameter input files (.prm files) """
+
+import sys
+import os
+import re
+import argparse
+
+__author__ = 'The authors of deal.II'
+__copyright__ = 'Copyright 2024, deal.II'
+__license__ = 'GNU GPL 2 or later'
+
+
+def get_parameter_value(parameters, name):
+    """ Given a dictionary of parameters with a structure as
+    the one created by read_parameter_file(), return the value
+    of the parameter with the given name. Returns None if the
+    parameter is not found.
+    """
+
+    if name in parameters:
+        return parameters[name]["value"]
+    else:
+        for entry in parameters:
+            if parameters[entry]["type"] == "subsection":
+                value = get_parameter_value(parameters[entry]["value"], name)
+                if value != None:
+                    return value
+
+    return None
+
+
+
+def set_parameter_value(parameters, name, value):
+    """ Given a dictionary of parameters with a structure as
+    the one created by read_parameter_file(), recursively search
+    through the subsections and set the value
+    of the first parameter with the given name to the given value.
+    Returns 0 if the parameter was found and set, 1 otherwise.
+    """
+
+    if name in parameters:
+        parameters[name]["value"] = value
+        return 0
+    else:
+        for entry in parameters:
+            if parameters[entry]["type"] == "subsection":
+                if set_parameter_value(parameters[entry]["value"], name, value) == 0:
+                    return 0
+
+    return 1
+
+
+def split_parameter_line(line):
+    """ Read a 'set parameter' line and extract the name, value, and format. """
+
+    equal_index = line.find('=')
+    # Determine number of additional spaces left of equal sign.
+    # We need to store these separately, because there might be two
+    # lines setting the same parameter, but different number of spaces. They
+    # should still map to the same parameter.
+    alignment_spaces = len(line[:equal_index]) - len(line[:equal_index].rstrip())
+
+    # skip initial word "set" in parameter name
+    words_left_of_equal_sign = line[:equal_index].replace("\\\n","").split()
+    param_name = words_left_of_equal_sign[1:]
+    param_name = ' '.join(param_name)
+
+    # strip spaces at end of value string, keep all inline comments
+    param_value = line[equal_index+1:].rstrip()
+    # if there is one or more spaces at the start of the string, remove one space.
+    # keep additional spaces, because they are part of the formatting.
+    # this way we can add one space when writing the file back and keep the formatting
+    if param_value.startswith(' ') and len(param_value) > 1:
+        param_value = param_value[1:]
+
+    return param_name, param_value, alignment_spaces
```

Comment:
It perhaps does not matter very much, but I think this all would have been easier if you had just applied a regex of the form `^ *set (.*?)( *)= *(.*?) *$` and then captured the three marked parts. You might have to deal with spaces actually being tabs as well.